syntax = "proto3";

// A node (like a block in unreal engine blueprint)

message BaseNode {
    int32 id = 1;
    string name = 2;
    NodeType type = 3;
    optional string description = 4;
    repeated IOpin pins = 5;
    string code = 6;

}

// Input and output pins
message IOpin {
    int32 id = 1;
    string name = 2;
    string type = 3;
    oneof value {
        string string_value = 4;
        int32 int_value = 5;
        float float_value = 6;
        bool bool_value = 7;
    }
}

// A enum node that stores the debug level

enum DebugLevel {
    DEBUG = 0;
    INFO = 1;
    WARNING = 2;
    ERROR = 3;
}

// A connection between two nodes

message Connection {
    int32 from_id = 1;
    int32 to_id = 2;
}

// A Node type that can be any direct inheritant of BaseNode

enum NodeType {
    BASE = 0;
    PRINT = 2;
    LOG = 3;
    CONDITION = 4;
    IF = 5;
}

// A node function
message NodeFunction {
    string name = 1;
    string description = 2;
    repeated IOpin inputs = 3;
    int32 first_node_id = 4;
}

// A graph that contains a list of nodes and a list of connections

message NodeGraph {
    repeated BaseNode nodes = 1;
    repeated Connection connections = 2;
}

